### PR TITLE
Add OLMo November 2024

### LINKED ISSUE #34496: Add support for OLMo November release
### Model description

An updated OLMo model will be released in November. The new model has a few small architecture changes compared to the existing model in transformers:

- RMSNorm is used instead of standard layer norm.
- Norm is applied to attention queries and keys.
- Norm is applied after attention/feedforward rather than before.

It would be great to have it integrated in transformers from day 1.

### Open source status

- [X] The model implementation is available
- [ ] The model weights are available

### Provide useful links for the implementation

OLMo code repo: https://github.com/allenai/OLMo
Authors include @2015aroras 
Weights are not yet available, will be made available at release.

### PR DESCRIPTION
**Pull request by 2015aroras**
# What does this PR do?

An updated OLMo model will be released in November. The new model has a few small architecture changes compared to the existing model in transformers:

- RMSNorm is used instead of standard layer norm.
- Norm is applied to attention queries and keys.
- Norm is applied after attention/feedforward rather than before.

The original PR https://github.com/huggingface/transformers/pull/34497 updated the OLMo implementation in transformers to support the November release. This PR instead adds a new model using the modular approach.

@ArthurZucker 

Fixes #34496 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by 2015aroras** [comment]
I tested this before we locked in the `Olmo1124` naming conventions. Will update once tested.

**Comment by 2015aroras** [comment]
Tests are passing, including slow ones. I've used a test HF hub repo ([shanearora/OLMo-7B-1124-hf](https://huggingface.co/shanearora/OLMo-7B-1124-hf)) since the official final model is not ready yet.

**Comment by 2015aroras** [comment]
PR checks were passing before I merged main again, and PR check failures relate to other models.

**Comment by 2015aroras** [comment]
@ArthurZucker Gentle ping

File: docs/source/en/model_doc/olmo_1124.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+⚠️ Note that this file is in Markdown but contain specific syntax for our doc-builder (similar to MDX) that may not be
+rendered properly in your Markdown viewer.
+
+-->
+
+# OLMo November 2024
+
+## Overview
+
+The OLMo November 2024 model was proposed in [<INSERT PAPER NAME HERE>](<INSERT PAPER LINK HERE>) by <INSERT AUTHORS HERE>.
```
Comment (ArthurZucker) [inline docs/source/en/model_doc/olmo_1124.md:21]: these would need to be filled

File: src/transformers/models/olmo_1124/__init__.py
Comment (ArthurZucker) [inline src/transformers/models/olmo_1124/__init__.py:1]: new init should look more like thisL https://github.com/huggingface/transformers/blob/aca9120a6e71b77edd86ff63a6cb0e3a998cb4af/src/transformers/models/albert/__init__.py#L4 
Should make it simpler 🤗

File: src/transformers/models/olmo_1124/modular_olmo_1124.py
Comment (ArthurZucker) [inline src/transformers/models/olmo_1124/modular_olmo_1124.py:1]: Very very well down, the modular is super simple, makes it easy to identify differences!

File: tests/models/olmo_1124/test_modeling_olmo_1124.py [hunk 2]
```diff
@@ -0,0 +1,467 @@
...
+    def test_export_static_cache(self):
+        if version.parse(torch.__version__) < version.parse("2.4.0"):
+            self.skipTest(reason="This test requires torch >= 2.4 to run.")
+
+        from transformers.integrations.executorch import (
+            TorchExportableModuleWithStaticCache,
+            convert_and_export_with_cache,
+        )
+
+        olmo_1124_model = "shanearora/OLMo-7B-1124-hf"
```
Comment (ArthurZucker) [inline tests/models/olmo_1124/test_modeling_olmo_1124.py:417]: is this the final checkpoint? 🤗

File: tests/models/olmo_1124/test_modeling_olmo_1124.py [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+            attn_implementation=attn_implementation,
+            generation_config=GenerationConfig(
+                use_cache=True,
+                cache_implementation=cache_implementation,
+                max_length=max_generation_length,
+                cache_config={
+                    "batch_size": batch_size,
+                    "max_cache_len": max_generation_length,
+                },
+            ),
```
Comment (ArthurZucker) [inline tests/models/olmo_1124/test_modeling_olmo_1124.py:446]: let's create it outside the call!

**Review by ArthurZucker — APPROVED** [review]
Looks marvellous thanks for your hard work, let's get this merged asap! 🤗 Left very small comments it's great. 
apologies again for the delay

File: tests/models/olmo_1124/test_modeling_olmo_1124.py (hunk 2 above)
Comment (2015aroras) [reply tests/models/olmo_1124/test_modeling_olmo_1124.py:417]: No, I just grabbed an intermediate checkpoint to use for the implementation. It's from pretty close to end of training.

We will upload the official final and intermediate checkpoints in an official HF Hub repo under the `allenai` org. Now that this PR is approved, I think we can start the uploading.

File: tests/models/olmo_1124/test_modeling_olmo_1124.py (hunk 3 above)
Comment (2015aroras) [reply tests/models/olmo_1124/test_modeling_olmo_1124.py]: [Create GenerationConfig outside of from_pretrained call](https://github.com/huggingface/transformers/pull/34551/commits/5b7cad9536f1156a62235a4978f97a68fd3737bf) This was auto-generated from `transformers-cli add-new-model-like`, but fixed anyways.

File: src/transformers/models/olmo_1124/__init__.py
Comment (2015aroras) [reply src/transformers/models/olmo_1124/__init__.py:1]: [f29dc50](https://github.com/huggingface/transformers/pull/34551/commits/f29dc5024c1feca23f1507f1f5a869c5a8a05c09) Done, though it took me a while to debug why it wasn't working. The simplified init requires `__all__` to be explicitly set: [3960e35](https://github.com/huggingface/transformers/pull/34551/commits/3960e3540087094ba1745178fe80b391a40c6f0b).

**Comment by ArthurZucker** [comment]
Something has gone wrong with the rebasing it seems 😓

**Comment by ArthurZucker** [comment]
We can merge once this is fixed!

**Comment by 2015aroras** [comment]
I'm just going to add something for the model card (better than the blank state it is, we can change it later).

File: docs/source/en/model_doc/olmo_1124.md [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+## Overview
+
+The OLMo November 2024 model is a successor of the OLMo model, which was proposed in
+[OLMo: Accelerating the Science of Language Models](https://arxiv.org/abs/2402.00838).
+
+ The architectural changes from the original OLMo model to this model are:
+
+- RMSNorm is used instead of standard layer norm.
+- Norm is applied to attention queries and keys.
+- Norm is applied after attention/feedforward layers rather than before.
```
Comment (ArthurZucker) [inline docs/source/en/model_doc/olmo_1124.md:28]: 💘 super clear, love this!

**Review by ArthurZucker — APPROVED** [review]
Merging! Thanks for the clean work 🤗

**Pull request merged**